src/backend/marsha/core/lti/utils.py

```diff
@@ -0,0 +1,144 @@
+        **filter_kwargs,
+    )
+
+
+def _get_or_create_playlist(model, lti):
+    # Check that the video does not already exist (possible in another playlist and/or
+    # on another consumer site if it is not portable or not ready).
+    try:
+        resource = model.objects.get(pk=lti.resource_id)
+    except model.DoesNotExist:
+        pass
+    else:
+        raise LTIException(
+            "The resource ID {!s} already exists but is not portable to your playlist ({!s}) "
+            "and/or consumer site ({!s}).".format(
+                resource.id, lti.context_id, lti.get_consumer_site().domain
+            )
+        )
+
+    playlist, _ = Playlist.objects.get_or_create(
+        lti_id=lti.context_id,
+        consumer_site=lti.get_consumer_site(),
+        defaults={"title": lti.context_title},
+    )
+
+    return playlist
+
+
+def get_or_create_video(lti):
+    """Get or create the video targeted by the LTI launch request.
+
+    Create the playlist if it does not pre-exist (it can only happen with consumer site scope
+    passports).
+
+    Parameters
+    ----------
+    lti : Type[LTI]
+
+    Raises
+    ------
+    LTIException
+        Exception raised if the request does not contain a context id (playlist).
+
+    Returns
+    -------
+    core.models.video.Video
+        The video instance targeted by the launch url or None.
+
+    """
+    # Make sure LTI verification has run successfully. It raises an LTIException otherwise.
+    lti.verify()
+    try:
+        video = _get_resource(Video, lti)
+    except Video.DoesNotExist:
+        if not lti.is_instructor:
+            return None
+    else:
+        return video
+
+    # Creating the video...
+    # - Get the playlist if it exists or create it
+    playlist = _get_or_create_playlist(Video, lti)
+
+    # Create the video, pointing to the file from the origin video if any
+    return Video.objects.create(
+        pk=lti.resource_id,
+        lti_id=lti.resource_link_id,
+        playlist=playlist,
+        upload_state=PENDING,
+        title=lti.resource_link_title,
+        show_download=lti.get_consumer_site().video_show_download_default,
+    )
+
+
+def get_or_create_file(lti):
+    """Get or create the file targeted by the LTI launch request.
+
+    Create the playlist if it does not pre-exist (it can only happen with consumer site scope
+    passports).
+
+    Parameters
+    ----------
+    lti : Type[LTI]
+
+    Raises
+    ------
+    LTIException
+        Exception raised if the request does not contain a context id (playlist).
+
+    Returns
+    -------
+    core.models.video.Video
+        The video instance targeted by the launch url or None.
+
+    """
+    # Make sure LTI verification has run successfully. It raises an LTIException otherwise.
+    lti.verify()
+    try:
+        file = _get_resource(File, lti)
+    except File.DoesNotExist:
+        if not lti.is_instructor:
+            return None
+    else:
+        return file
```

Comment:
file is a reserved word in Python. You should find another name.

Reply:
:+1: